### DISCUSSION TITLE
SqlQuery Expression custom select include t0 before foreign table alias

### DISCUSSION THREAD

**Discussion by minhhungit** [serenity-is/Serenity#5621] [Help]
This happen at .NET 5 version, **.net core 3.1 version work correct**

In my row, I have a field from foreign table using LeftJoin

![image](https://user-images.githubusercontent.com/2279508/114110883-c213e480-9902-11eb-8b35-62a690a982f2.png)

and its expression: `jReleaseFormat.[ReleaseFormatDescription]`
![image](https://user-images.githubusercontent.com/2279508/114110736-677a8880-9902-11eb-9528-2db415f02d43.png)

When I write custom query it generated `t0` before `jReleaseFormat` alias in TSQL 

>t0_jReleaseFormat.[ReleaseFormatDescription]

![image](https://user-images.githubusercontent.com/2279508/114111871-436c7680-9905-11eb-8824-70a8aebf2f99.png)

Or :

![image](https://user-images.githubusercontent.com/2279508/114110683-53cf2200-9902-11eb-846a-af3953e7ff50.png)

**Comment by VictorTomaili** [comment]
So what the error is you faced? It's join table field, and joins with t0 table. if you have more than one same table,  how can you determine differences between "jReleaseFormat"? Also you are explicitly use T0 aliased fields with query customization. 
You didn't point to a specific exception or bug.

**Reply by VictorTomaili** [reply]
@minhhungit can you show how you define t0 variable too?

**Reply by minhhungit** [reply]
Sorry I forgot to add error message, here it is:

```
The multi-part identifier "t0_jReleaseFormat.ReleaseFormatDescription" could not be bound.
The multi-part identifier "t0_jReleaseProfile.ReleaseProfileName" could not be bound.
The multi-part identifier "t0_jReleaseStatus.ReleaseStatusDescription" could not be bound.
The multi-part identifier "t0_jProductReleaseType.ProductReleaseType" could not be bound.
```

This is sql which is generated:
```sql
SELECT 
  T0.[ProductReleaseID] AS [ProductReleaseId],  
  T0.[MasterReleaseID] AS [MasterReleaseId], 
  // ...
  // ...
  T0.[InsertDate] AS [InsertDate], 
  T0.[UpdateUserId] AS [UpdateUserId], 
  T0.[UpdateDate] AS [UpdateDate], 
  jParentalWarningType.[ParentalWarningType] AS [ParentalWarningType], 
  jProductReleaseVersion.[ProductReleaseVersion] AS [ProductReleaseVersion], 
  jReleaseProfile.[ReleaseProfileName] AS [ReleaseProfileName], 
  // ...
  // ...
  jReleaseFormat.[ReleaseFormatDescription] AS [ReleaseFormatDescription], 
  jProductReleaseType.[ProductReleaseType] AS [ProductReleaseType], 
  jReleaseStatus.[ReleaseStatusDescription] AS [ReleaseStatusDescription], 
  jGenre.[AppleGenreCode] AS [GenreAppleGenreCode], 
  jGenre.[GenreDescription] AS [GenreGenreDescription], 
  jMasterReleaseType.[MasterReleaseTypeDescription] AS [MasterReleaseTypeDescription], 
  jMasterReleaseStatus.[ReleaseStatusDescription] AS [MasterReleaseStatusDescription], 
  t0.[ProductReleaseID] AS [ProductReleaseId], 
  t0.[MasterReleaseID] AS [MasterReleaseId], 
  t0_jReleaseFormat.[ReleaseFormatDescription] AS [ReleaseFormatDescription], 
  t0_jReleaseProfile.[ReleaseProfileName] AS [ReleaseProfileName], 
  t0_jReleaseStatus.[ReleaseStatusDescription] AS [ReleaseStatusDescription], 
  t0_jProductReleaseType.[ProductReleaseType] AS [ProductReleaseType] 
FROM 
  [ProductReleases] T0 
  LEFT JOIN [dbo].[ReleaseFormats] jReleaseFormat ON (
    jReleaseFormat.[ReleaseFormatID] = T0.[ReleaseFormatID]
  ) 
  LEFT JOIN [dbo].[ReleaseProfiles] jReleaseProfile ON (
    jReleaseProfile.[ReleaseProfileID] = T0.[ReleaseProfileID]
  ) 
  // ...
WHERE 
  (t0.[MasterReleaseID] = 50)
```

Error at 4 last fields

Agree that t0 is alias of the main table but those field have Expression attributes, those fields are from foreign table, generate should use exactly expression string

Expected is it should not auto add "t0_" string before custom expression

**Reply by minhhungit** [reply]
this is how I define t0:

```csharp
var t0 = ProductReleaseRow.Fields.As("t0");
```

**Reply by minhhungit** [reply]
If I update alias string in above command like this:

>var t0 = ProductReleaseRow.Fields.As("xxxx");

Then it generate sql :

![image](https://user-images.githubusercontent.com/2279508/114146340-ad0a7600-9941-11eb-9c30-03a87b9eda32.png)

**Reply by VictorTomaili** [reply]
Just remove `As` and use `var t0 = ProductReleaseRow.Fields;` instead.

**Reply by minhhungit** [reply]
Remove `As()` works but I feel that it is just workarounds, I still think that if we defined [Expression], it just should use Expression string instead

**Reply by VictorTomaili** [reply]
Nope It's not a workaround, you cannot use T0 as alias field. you need to use just fields as is. Default alias is already T0.

**Reply by minhhungit** [reply]
yes, I know default alias is t0

But if I change alias to another name it should use new name. That's clear.

Also if I use expression attribute it should use expression string without modifying it. There is no reason we have to modify expression, let expession does its work, expression does not need to care about alias I think, **it will be confusing when sometime we use or not use alias for expression**

Although I will try to remove `As()` and update a lot queries in many projects to fix but I still trust that it is a bug LOL 😅

**Reply by VictorTomaili** [reply]
> Also if I use expression attribute it should use expression string without modifying it. There is no reason we have to modify expression, let expession does its work, expression does not need to care about alias I think

Imagine that, you have [dbo].[Category] table what joins self for category tree. and it's has join to other table like [dbo].[CategoryTags] with expression.
If joins didn't modify the expression prefix, how you get subcategory tags list or parent tags list?

- for self 
`CategoryTags.Tag`
- for join table
`T1_CategoryTags.Tag`

> it will be confusing when sometime we use or not use alias for expression

It's not confusing. When you need to modify query for custom result, you need to ensure main table fields and main table expression fields must be without alias. Or must include alias in [From](https://serenity.is/Docs/api/dotnet/Serenity.Net.Data/Serenity.Data/SqlQuery/From) method as parameter. Other joins must have alias.

**Reply by minhhungit** [reply]
I am not sure which case you are refer to, JOIN on row class or create SqlQuery, I guess you want join on row.cs

I tried to create sample here, you can get tag list of parent cateory or child category

![image](https://user-images.githubusercontent.com/2279508/114219042-6134ec80-9994-11eb-8cd3-86f51886d8da.png)

this is data

```sql
DECLARE @category table (
	ID INT,
	Name NVARCHAR(100),
	ParentCategoryId INT
)

INSERT INTO @category
SELECT 1, 'Abc', NULL

INSERT INTO @category
SELECT 2, 'Xyz', NULL

INSERT INTO @category
SELECT 3, 'abc sub 1', 1

DECLARE @categoryTags table (
	ID INT,
	CategoryId INT,
	Name NVARCHAR(100)
)

INSERT INTO @categoryTags
SELECT 1, 1, 'abc tag 1'

INSERT INTO @categoryTags
SELECT 1, 2, 'xyz tag 1'

INSERT INTO @categoryTags
SELECT 1, 3, 'abc sub tag 1'

INSERT INTO @categoryTags
SELECT 1, 3, 'abc sub tag 2'

SELECT 
  T0.Id AS [CategoryId],
  T0.[Name] AS [Name],
  T0.ParentCategoryId,
  jParent.[Name] AS [ParentCategoryName],
  CASE WHEN T0.ParentCategoryId IS NULL THEN pct.Name ELSE cct.Name END TagName
FROM @category T0 
LEFT JOIN @category jParent ON (jParent.Id = T0.ParentCategoryId) 
LEFT JOIN @categoryTags as pct on pct.CategoryId = t0.Id AND T0.ParentCategoryId IS NULL
LEFT JOIN @categoryTags as cct on cct.CategoryId = t0.Id AND T0.ParentCategoryId IS NOT NULL
ORDER BY T0.[Name], T0.Id
go
```

Because I am not sure how you want to show tag data on grid, this case is not clear so I don't have exactly answer, but by using JOIN like my image it should work, it just has problem about duplicate ID on slickgrid

Btw I guess for this case we normally need to use sql view instead

**Reply by minhhungit** [reply]
Here is real data I created using dbo.BusinessUnits (pretend it's category) and dbo.Contact (category tag)

Sorry too lazy to create new table and generate classes :)

![image](https://user-images.githubusercontent.com/2279508/114220337-0ef4cb00-9996-11eb-8497-9dad6b37658b.png)


and full code of row.cs

```csharp
using Serenity.ComponentModel;
using Serenity.Data;
using Serenity.Data.Mapping;
using System;
using System.ComponentModel;


namespace StartSharpVer5.Organization.Entities
{
    [ConnectionKey("Default"), Module("Organization"), TableName("BusinessUnits")]
    [DisplayName("Business Units"), InstanceName("Business Unit")]
    [ReadPermission(PermissionKeys.BusinessUnits.Management)]
    [ModifyPermission(PermissionKeys.BusinessUnits.Management)]
    [LookupScript(Permission = PermissionKeys.General)]
    [LeftJoin("pc", "dbo.Contacts", "t0.UnitId = pc.IdentityNo AND T0.[ParentUnitId] IS NULL")]
    [LeftJoin("cc", "dbo.Contacts", "t0.UnitId = cc.IdentityNo AND T0.[ParentUnitId] IS NOT NULL")]
    public sealed class BusinessUnitRow : Row<BusinessUnitRow.RowFields>, IIdRow, INameRow
    {
        [DisplayName("Unit Id"), Identity, IdProperty]
        public Int32? UnitId
        {
            get => fields.UnitId[this];
            set => fields.UnitId[this] = value;
        }

        [DisplayName("Name"), Size(100), NotNull, QuickSearch, NameProperty]
        public String Name
        {
            get => fields.Name[this];
            set => fields.Name[this] = value;
        }

        [DisplayName("Parent Unit"), ForeignKey("BusinessUnits", "UnitId"), LeftJoin("jParentUnit"), TextualField("ParentUnitName")]
        [LookupInclude, BusinessUnitEditor]
        public Int32? ParentUnitId
        {
            get => fields.ParentUnitId[this];
            set => fields.ParentUnitId[this] = value;
        }

        [DisplayName("Parent Unit Name"), Expression("jParentUnit.[Name]")]
        public String ParentUnitName
        {
            get => fields.ParentUnitName[this];
            set => fields.ParentUnitName[this] = value;
        }

        [DisplayName("Parent Unit Parent Unit Id"), Expression("jParentUnit.[ParentUnitId]")]
        public Int32? ParentUnitParentUnitId
        {
            get => fields.ParentUnitParentUnitId[this];
            set => fields.ParentUnitParentUnitId[this] = value;
        }

        [DisplayName("Title"), Expression("CASE WHEN T0.[ParentUnitId] IS NULL THEN pc.Title ELSE cc.Title END")]
        public String Title
        {
            get => fields.Title[this];
            set => fields.Title[this] = value;
        }

        public BusinessUnitRow()
        {
        }

        public BusinessUnitRow(RowFields fields)
            : base(fields)
        {
        }

        public class RowFields : RowFieldsBase
        {
            public Int32Field UnitId;
            public StringField Name;
            public Int32Field ParentUnitId;

            public StringField ParentUnitName;
            public Int32Field ParentUnitParentUnitId;
            public StringField Title;
        }
    }
}
````

**Reply by minhhungit** [reply]
>It's not confusing. When you need to modify query for custom result, you need to ensure main table fields and main table expression fields must be without alias. Or must include alias in From method as parameter. Other joins must have alias.

maybe you misunderstood, I mean with some Expression serenity will include custom alias before my expression, I see normally it is "explicit" field for example with `[Expression(jProduct.MyFieldOnForignTable)]` it will generate code:

>xxx_jProduct.MyFieldOnForignTable

but for some expression like [Expression(1+ 1)] or [Expression(case when...))] it does not include custom `xxx` alias into these expression 

That why I said it's quite confusing

**Reply by minhhungit** [reply]
Btw, back to this command `var t0 = ProductReleaseRow.Fields.As("xxx");`

I did not join with the alias, I just use it to get its field names and its expression I defined in row. I think if my query does not join with the alias, there is no reason Serenity use `xxx` and modify my expression right ?

So maybe we need to check join before using alias
